src/demo-app/data-table/person-data-source.ts

```diff
@@ -1,29 +1,58 @@
-import {CollectionViewer, DataSource} from '@angular/material';
+import {CollectionViewer, DataSource, MdPaginator, PageEvent} from '@angular/material';
 import {Observable} from 'rxjs/Observable';
 import {PeopleDatabase, UserData} from './people-database';
+import {BehaviorSubject} from 'rxjs/BehaviorSubject';
 
 export class PersonDataSource extends DataSource<any> {
+  /** Data that should be displayed by the table. */
+  _displayData = new BehaviorSubject<UserData[]>([]);
+
+  /** Cached data provided by the display data. */
   _renderedData: any[] = [];
 
-  constructor(private _peopleDatabase: PeopleDatabase) {
+  constructor(private _peopleDatabase: PeopleDatabase,
+              private _paginator: MdPaginator) {
     super();
+
+    // Subscribe to page changes by clearing the cached data and
+    // determining the updated display data.
+    this._paginator.page.subscribe(() => {
+      this._renderedData = [];
+      this.updateDisplayData();
+    });
   }
 
   connect(collectionViewer: CollectionViewer): Observable<UserData[]> {
-    return collectionViewer.viewChange.map((view: {start: number, end: number}) => {
-      // Set the rendered rows length to the virtual page size. Fill in the data provided
-      // from the index start until the end index or pagination size, whichever is smaller.
-      this._renderedData.length = this._peopleDatabase.data.length;
+    this.updateDisplayData();
 
-      const buffer = 20;
-      let rangeStart = Math.max(0, view.start - buffer);
-      let rangeEnd = Math.min(this._peopleDatabase.data.length, view.end + buffer);
+    const streams = [collectionViewer.viewChange, this._displayData];
+    return Observable.combineLatest(streams)
+        .map((results: [{start: number, end: number}, UserData[]]) => {
+          const [view, data] = results;
 
-      for (let i = rangeStart; i < rangeEnd; i++) {
-        this._renderedData[i] = this._peopleDatabase.data[i];
-      }
+          // Set the rendered rows length to the virtual page size. Fill in the data provided
+          // from the index start until the end index or pagination size, whichever is smaller.
+          this._renderedData.length = data.length;
 
-      return this._renderedData;
-    });
+          const buffer = 20;
+          let rangeStart = Math.max(0, view.start - buffer);
+          let rangeEnd = Math.min(data.length, view.end + buffer);
+
+          for (let i = rangeStart; i < rangeEnd; i++) {
+            this._renderedData[i] = data[i];
```

Comment:
This is because I'll want renderedData to be a virtual buffer that fills in over time. E.g. the table says it wants rows 0-10 and then the user scrolls to the end and asks for 90-100. Rendered data should be an array that contains 100 items that are all undefined except for 0-10 and 90-100